src/sandstorm/run-bundle.c++

```diff
@@ -1556,36 +1563,65 @@ private:
+      args.add("/bin/mongod");
+      args.add("--fork");
+      args.add("--bind_ip");
+      args.add("127.0.0.1");
+      args.add("--port");
+      args.add(kj::str(config.mongoPort).cStr());
+      args.add("--dbpath");
+      args.add("/var/mongo");
+      args.add("--logpath");
+      args.add("/var/log/mongo.log");
+      args.add("--pidfilepath");
+      args.add("/var/pid/mongo.pid");
+      args.add("--auth");
+      args.add("--nohttpinterface");
+      args.add("--noprealloc");
+      args.add("--nopreallocj");
+      args.add("--smallfiles");
+      args.add("--replSet");
+      args.add("ssrs");
+      args.add("--oplogSize");
+      args.add("16");
+
+      if (doMongoRepair) {
+        args.add("--repair");
+      }
+
+      args.add(nullptr);
+
+      context.warning(kj::str("MONGO ARGS: ", args));
+
+      KJ_SYSCALL(execv(args[0], const_cast<char**>(args.begin())));
       KJ_UNREACHABLE;
     });
 
     // Wait for mongod to return, meaning the database is up.  Then get its real pid via the
     // pidfile.
-    process.waitForSuccess();
-
-    // Even after the startup command exits, MongoDB takes exactly two seconds to elect itself as
-    // master of the repl set (of which it is the only damned member). Unforutnately, if Node
-    // connects during this time, it fails, sometimes without actually exiting, leaving the entire
-    // server hosed. It appears that this always takes exactly two seconds from startup, since
-    // MongoDB does some sort of heartbeat every second where it checks the replset status, and it
-    // takes three of these for the election to complete, and the first of the three happens
-    // immediately on startup, meaning the last one is two seconds in. So, we'll sleep for 3
-    // seconds to be safe.
-    // TODO(cleanup): There must be a better way...
-    int n = 3;
-    while (n > 0) n = sleep(n);
+    int exitCode = process.waitForExit();
+
+    if (exitCode == 0) {
+      // Even after the startup command exits, MongoDB takes exactly two seconds to elect itself as
+      // master of the repl set (of which it is the only damned member). Unforutnately, if Node
+      // connects during this time, it fails, sometimes without actually exiting, leaving the entire
+      // server hosed. It appears that this always takes exactly two seconds from startup, since
+      // MongoDB does some sort of heartbeat every second where it checks the replset status, and it
+      // takes three of these for the election to complete, and the first of the three happens
+      // immediately on startup, meaning the last one is two seconds in. So, we'll sleep for 3
+      // seconds to be safe.
+      // TODO(cleanup): There must be a better way...
+      int n = 3;
+      while (n > 0) n = sleep(n);
+    } else {
+      context.warning(kj::str("** Mongo exited with exitCode: ", exitCode));
+    }
 
     return KJ_ASSERT_NONNULL(parseUInt(trim(readAll("/var/pid/mongo.pid")), 10));
```

Comment:
+1, doing the repair synchronously strikes me as very reasonable, given that the frontend will break in half during that time.